docs/tech_specs/screens_by_alert/tech_spec.md

```diff
@@ -0,0 +1,187 @@
+# Screens by alert real-time data source - tech spec
+
+## Need
+
+Screenplay (and maybe other services, in the future) needs real-time insight into which screens are
+displaying info about a given service alert.
+
+## High-level architecture
+
+To store and provide this new data, we will add a new caching layer to the Screens application.
+
+To keep the cache in one shared location, we will use memcached rather than an ETS table or GenServer/other stateful process running on each server instance. This follows the precedent of the V3 API, which uses memcached to keep track of per-user request counts and apply throttling appropriately no matter how many instances it's scaled to.
+
+### Data structure
+
+Cached values will be as follows, defined as Elixir typespecs:
+```ex
+@type alert_id :: String.t()
+@type screen_id :: String.t()
+@type timestamp :: integer()  # we'll use System.monotonic_time/0 to produce these
+@type timestamped_screen_id :: {screen_id, timestamp}
+
+@type cache :: %{
+  "screens_by_alert." <> alert_id => list(timestamped_screen_id),
+  # metadata to make the "self-refresh" mechanism possible
+  "screens_last_updated." <> screen_id => timestamp
+}
+```
+
+TTL for `screens_by_alert.*` cache keys will be a number just over the "self-refresh" interval, so that in the case of a screen client going down we don't inaccurately report it as showing no alerts for a short period before the "self-refresh" mechanism kicks in.
+
+TTL for `screens_last_updated.*` keys will be something large, like 1 day. These values should only expire when a screen is deleted from the config and
+the self-refresh logic stops caring about it.
```

Comment:
TTL: I first assumed that this was the drop-dead expiration of a value, but really it seems like this is the "time until we run some logic to determine whether the value should expire". Is that right?

Reply:
In our case it's kind of both unfortunately 😬

In the text you highlighted for this comment, it's the former ("the drop-dead expiration of a value"). We'll specify the TTL when updating these keys via memcached commands, and memcached will carry out their expirations accordingly.

For other values in the cache, specifically each screen ID in the list under an alert key, _we_ need to implement the TTL ourselves because memcached is totally agnostic of the internal structure of data you store under a key. So in that case, "we run some logic to determine whether the value should expire" is correct.

Reply:
A minor thing, but does the TTL need to be very large? If a screen is deleted from the config, then it seems like the self-refresh GenServer would realize that in a minute or so at which point it would stop attempting self-refreshes. It seems like the `screen_last_updated` key could be dropped immediately or soon after that.

Reply:
I think it's possible to cover that base while keeping the TTL large. Because the TTL is a "check if we need to self-refresh" timer, we could do a check against `State.screens()` to check that the screen still exists. If it's still there, do a self-refresh if we are past whatever the "TTL" is. 

(An out-of-scope observation, we might also be able to use this to clean up `places_and_screens.json`)

Reply:
> Because the TTL is a "check if we need to self-refresh" timer

🤔 Hold up... I think the "large TTL" referenced here in the doc is one thing (the timeline by which memecached will expire un-updated screens) and the "check if we need to self-refresh" is a separate timer that we manage in a GenServer. Right? So Paul K's suggestion makes sense to me -- the `screen_last_updated` TTL can be a couple minutes, since the screen probably would have refreshed a few times by then.

Reply:
Yeah, I think you're right! I left a longer gap mostly because I don't yet have a good sense of how long the self-refresh will take**, on average, and because the GenServer will already ignore these keys anyway. The TTL is there just to have them stop taking up space in the cache after a while.

**TBH it will probably be decently fast—the prod server responds to screen data requests in about 100-200ms, and I'm sure the actual computation time is significantly smaller than that. Multiply that by how many screen clients we expect to be down at a given moment, on average—like 10 max? And that's roughly 1-2 seconds of computation time.

Reply:
hmm so there's little to no risk of having old data in the cache then right? Initially I was following a train of thought that led me to wonder if it's plausible that future alerts-related logic in Screenplay would still think that a screen exists even though the config for it was deleted. 

I think it should be fine though if I'm understanding the spec correctly since the deleted screen_id would naturally be removed from the alert_id's list by virtue of the timestamps expiring. Does that check out @jzimbel-mbta ? 🤔 

Ohh and maybe this concern is more something that would be addressed by eventually not having to manually update the places and screens list anymore